_drafts/diving-into-jtag-part6.md

```diff
@@ -0,0 +1,373 @@
+</p>
+
+Another protection option is suitable when there is no JTAG connector on the board at all, and the
+JTAG pins are scattered across the board as test points.
+
+<p align="center">
+ <img width="100%" src="{% img_url jtag-part6/board-level-protection-2.png %}" alt="The Board-level protection example 2" />
+</p>
+
+However, as you well understand, these methods provide only the most basic protection, as soldering a
+JTAG connector is not a significant effort, nor is finding the necessary JTAG pins on the microcontroller,
+especially when its part number is written on the package, as shown in the photo. But even if the
+microcontroller is unknown, there are ways to scan the pins of the microcontroller to discover the necessary
+JTAG pins (I will discuss these methods in more detail later).
+
+## Chip-Level
+
+Debug interfaces are convenient during device development and debugging, but
+it is good practice to disable them in the release version of the device/firmware.
+There are configurable bits in the microcontroller's non-volatile memory to disable
+certain debug interface functions. The flexibility of configuring protection
+features depends on the specific core and microcontroller model.
+
+For instance, access to flash memory can be disabled. Some microcontroller models
+support completely disabling the debug interface. The exact configuration of possible
+protection features can be found in the microcontroller's documentation. An example is
+the Flash Readout Protection (RDP) technology in STM32 microcontrollers.
+This technology allows protecting the contents of the microcontroller's embedded
+flash memory from being read through the debug interface.
+
+<p align="center">
+ <img width="100%" src="{% img_url jtag-part6/rdp-levels.png %}" alt="The RDP Levels" />
+</p>
+
+RDP has three levels of protection (0, 1, and 2):
+
+ - **Level 0** (default RDP level)
+The flash memory is fully open, and all memory operations are possible in all boot
+configurations (debug features, boot from RAM, boot from system memory bootloader,
+boot from flash memory). There is no protection in this configuration mode that is
+appropriate only for development and debug.
+- **Level 1**
+Flash memory accesses (read, erase, program), or SRAM2 accesses via debug features
+(such as serial-wire or JTAG) are forbidden, even while booting from SRAM or system
+memory bootloader. In these cases, any read request to the protected region generates
+a bus error. However, when booting from flash memory, accesses to both flash memory
+and to SRAM2 (from user code) are allowed.
+- **Level 2**
+All protections provided in Level 1 are active, and the MCU is fully protected. The
+RDP option byte and all other option bytes are frozen, and can no longer be modified.
+The JTAG, SWV (single-wire viewer), ETM, and boundary scan are all disabled.
+
+RDP can always be leveled up. A level regression is possible with the following
```

Comment:
For those unfamiliar with this protection mechanism, could you describe the situations in which you would level up and down? Without having read about it before, I'm thinking the main use case is if you want to re-program a device that has a RDP level set, e.g. in the case of a production device device that has been returned and you are manually fixing it.

Reply:
Fixed